### ISSUE TITLE
No float 16 on esp32?

### ISSUE THREAD

**Issue by 963io** [peterzuger/bitstruct-micropython#3]
Hey! Not sure if I write in the right place, but it seems f16 doesn't work on esp32, only f32. Does this means platform doesn't support it or library doesn't implement it for the platform?

**Comment by peterzuger** [comment]
Hi, Yes this is the right place.

I had never really tested f16 support until now. It would not have worked until now since there where some mistakes in the code, they should be fixed now.

Again I'm sorry I can't verify this on esp32 yet, but on stm32 f16 support can be enabled with a define and an additional compiler flag:

`-D__STDC_WANT_IEC_60559_TYPES_EXT__=1 -mfp16-format=ieee`
The first enables _Float16(and other) support in the glibc library, the second one sets the _Float16 format in the compiler.

The ESP platform uses gcc as well so the same should work there too.

**Comment by 963io** [comment]
Hi @peterzuger ! Thank you so much for quick and detailed reply. I will test it with new parameters on my esp32 and get back to you.

Also I have a question, f16 packing and then unpacking produces varied results at the end, after point. Sure I round it, but my question is this due to nature of fpu operations (or to be more precise float 16 fpu operations), or because as you mentioned there was some code needed fixing? (Just sanity check here ;)

**Comment by peterzuger** [comment]
Yes, this has to do with the nature of floating point numbers, some numbers just cannot be represented.

With 16bit ieee floats for example 1.01 or 1.02 cannot be represented and so it will be rounded to an existing ieee 16bit float:
```
>>> bitstruct.unpack("f16", bitstruct.pack("f16", 1.01))
(1.009766,)
>>> bitstruct.unpack("f16", bitstruct.pack("f16", 1.02))
(1.019531,)
```


But when the input is a number that can be represented the output will always be the same:
```
bitstruct.unpack("f16", bitstruct.pack("f16", 1.009766))
(1.009766,)
>>> bitstruct.unpack("f16", bitstruct.pack("f16", 1.019531))
(1.019531,)
```

**Comment by 963io** [comment]
Nope, it didn't worked out. Still says this strange phrase...
`Error: expected float size of 32, or 64 bits (got 16)`

Code sample:
`bitstruct.pack('f16',round(random.uniform(-180,180),3))`

**Comment by peterzuger** [comment]
This means that the support for `_Float16` was not detected. The detection is done with this code [here](https://github.com/peterzuger/bitstruct-micropython/blob/4ebb0111fb9b1d516ef877d217100a63c57bbbad/bitstruct/bitstruct.c#L277).

The flags where probably not set correctly, you could set the first one like you did with the enable: [issue 2](https://github.com/peterzuger/bitstruct-micropython/issues/2#issuecomment-1100598425):
```
#define __STDC_WANT_IEC_60559_TYPES_EXT__
```

I don't know about the second flag, you have to make sure it is passed to gcc by the `idf.py` tool.


Reading about `idf.py` [here](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/build-system.html?highlight=rename#advanced-commands) you probably have to call it with reconfigure like this to set the define from the commandline:

```
idf.py -D__STDC_WANT_IEC_60559_TYPES_EXT__=1 reconfigure
```

The second flag probably has to be set in `CMakeLists.txt` (maybe via `idf.py menuconfig`).

This issue from esp-idf seems related: https://github.com/espressif/esp-idf/issues/8021#issuecomment-986889197

**Comment by 963io** [comment]
Well, I found out I can only enable modules or definitions in py/mpconfig.h. All my tries to do it in sources, in micropython.cmake, micropython.mk, CMakeLists.txt, and even in standalone make file (make -f makefile) failed. Everything would compile, but nothing would import.

![2022-04-20 17_30_12-Micropython  Running  - Oracle VM VirtualBox](https://user-images.githubusercontent.com/40085574/164254448-92d1b883-3575-4834-abb4-cc33d27bdc7a.png)

So, again, just for the sake of it, I tried adding these 2 variables in every aforementioned place. make clean, make -f makefile and flash. Same results .. :(
